win32/src/file_win32.c

```diff
@@ -60,4 +41,472 @@ static void on_file_io_complete_win32(
+                            result = false;
+                        }
+                        if (!result)
+                        {
+                            CloseThreadpoolCleanupGroup(handle->ptp_cleanup_group);
+                        }
+                    }
+                    if (!result)
+                    {
+                        DestroyThreadpoolEnvironment(&handle->cbe);
+                    }
+                }
+            }
+            /*Codes_SRS_FILE_43_034: [ If there are any failures, file_create shall fail and return NULL. ]*/
+            if (!result)
+            {
+                free(handle);
+            }
+        }
+    }
+    /*Codes_SRS_FILE_WIN32_43_008: [ If there are any failures, file_create shall return NULL.]*/
+    if (!result)
+    {
+        return NULL;
+    }
+
+    /*Codes_SRS_FILE_43_038: [ file_create shall register user_report_fault_callback with argument user_report_fault_context as the callback function to be called when the callback specified by the user for a specific asynchronous operation cannot be called. ]*/
+    handle->user_report_fault_callback = user_report_fault_callback;
+    handle->user_report_fault_context = user_report_fault_context;
+    /*Codes_SRS_FILE_WIN32_43_009: [ file_create shall succeed and return a non - NULL value.]*/
+    return handle;
+}
+
+
+IMPLEMENT_MOCKABLE_FUNCTION(, void, file_destroy, FILE_HANDLE, handle)
+{
+    /*Codes_SRS_FILE_43_005: [ If handle is NULL, file_destroy shall return. ]*/
+    /*Codes_SRS_FILE_WIN32_43_049: [ If handle is NULL, file_destroy shall return.]*/
+    if (handle != NULL)
+    {
+        /*Codes_SRS_FILE_43_006: [ file_destroy shall wait for all pending I/O operations to complete. ]*/
+        /*Codes_SRS_FILE_WIN32_43_011: [ file_destroy shall wait for all I/O to complete by calling WaitForThreadpoolIoCallbacks. ]*/
+        WaitForThreadpoolIoCallbacks(handle->ptp_io, FALSE);
+        /*Codes_SRS_FILE_WIN32_43_012: [ file_destroy shall close the cleanup group by calling CloseThreadpoolCleanupGroup. ]*/
+        CloseThreadpoolCleanupGroup(handle->ptp_cleanup_group);
+        /*Codes_SRS_FILE_WIN32_43_013: [ file_destroy shall destroy the environment by calling DestroyThreadpoolEnvironment. ]*/
+        DestroyThreadpoolEnvironment(&(handle->cbe));
+        /*Codes_SRS_FILE_43_007: [ file_destroy shall close the file handle handle. ]*/
+        /*Codes_SRS_FILE_WIN32_43_016: [ file_destroy shall call CloseHandle on handle->h_file ]*/
+        if (!CloseHandle(handle->h_file))
+        {
+            LogError("failure in CloseHandle");
+        }
+        /*Codes_SRS_FILE_WIN32_43_015: [ file_destroy shall close the threadpool IO by calling CloseThreadPoolIo. ]*/
+        CloseThreadpoolIo(handle->ptp_io);
+        /*Codes_SRS_FILE_WIN32_43_042: [ file_destroy shall free the handle.]*/
+        free(handle);
+    }
+    else
+    {
+        LogError("Invalid argument to file_destroy: FILE_HANDLE=%p", handle);
+    }

```

Comment:
Style: typically errors always go first (if is null print error else do stuff). Good for consistency/readability in the code-base